apps/studio/prisma/scripts/collection-tags/index.ts

```diff
@@ -0,0 +1,275 @@
+      (resource) => (resource.content.page as any).tags as LegacyTag[],
+    )
+
+    const baseTags: {
+      categories: Set<string>
+      mappings: Record<string, Set<string>>
+    } = {
+      categories: new Set(),
+      // NOTE: mappings denotes the mapping of category -> options
+      mappings: {},
+    }
+
+    const collatedTags = tags.reduce((prevTags, curTags) => {
+      curTags.forEach((tag) => {
+        prevTags.categories.add(tag.category)
+        tag.selected.forEach((value) => {
+          let prevCategorySet = prevTags.mappings[tag.category]
+          if (!prevCategorySet) {
+            prevTags.mappings[tag.category] = new Set()
+            prevCategorySet = prevTags.mappings[tag.category]
+          }
+
+          prevCategorySet!.add(value)
+        })
+      })
+
+      return prevTags
+    }, baseTags)
+
+    const { labelToId, tagCategories } = migrateTags(collatedTags)
+    const indexPage = await db
+      .selectFrom("Resource")
+      .where("type", "=", "IndexPage")
+      .where("parentId", "=", id)
+      .selectAll()
+      // NOTE: assumption - every collection has an index page
+      // we should always run the other migration `addCollectionIndexPage` first
+      .executeTakeFirstOrThrow()
+
+    // NOTE: we need to do 2 things in this `tx`
+    // Step 1: write to the index page with the newly generated tags
+    // Step 2: write the updated mappings to each individual page
+    await db.transaction().execute(async (tx) => {
+      const indexPageBlob = await getBlobOfResource({
+        tx,
+        resourceId: indexPage.id,
+      })
+
+      await updateBlobById(tx, {
+        pageId: Number(indexPage.id),
+        siteId: indexPage.siteId,
+        content: {
+          ...indexPageBlob.content,
+          page: { ...indexPageBlob.content.page, tags: tagCategories },
+        },
+      })
+
+      for (const resource of resourcesWithTags) {
+        const blob = resource.content
+        const tagsOfResource = (resource.content.page as any)
+          .tags as LegacyTag[]
+        // NOTE: we only update the `page.tags` here
+        // since the rest of the content is not changed
+        const updatedContent = {
+          ...blob,
+          page: {
+            ...blob.page,
+            // NOTE: we CANNOT reuse `tags` because our existing tags use it...
+            // If we replace the existing labels because it'll cause the sites to have the uuid
+            // on the next rebuild due to this feature not being released.
+            // We also cannot add on this label,
+            // because it will hinder deletion.
+            tagged: tagsOfResource
+              .flatMap(({ selected }) =>
+                selected.map((label) => labelToId[label]),
+              )
+              .filter((v) => !!v),
+          },
+        }
+
+        await updateBlobById(tx, {
+          pageId: Number(resource.id),
+          siteId: indexPage.siteId,
+          content: updatedContent,
+        })
```

Comment:
question: hmm we are not doing a publish, wouldn't that mean that the collection can be in a potentially weird state whereby some pages have the new tags and some still live on the old one?

Understand that we are using a new key `tagged` but not sure if we have plans to do a forceful migration to remove all the old tags

Reply:
Discussed offline, we should ideally do an update to the latest published version even for pages with an existing draft blob, so sites can build with the newer tags.